### PR TITLE
ZTF sentinel service

### PR DESCRIPTION
**Pull request by dmitryduev**
This PR implements a containerized Tails sentinel service for ZTF.

The sentinel:
- Managed with the `sentinel.py` utility.
- Implemented as a containerized service, orchestrated with `docker-compose`.
- Monitors the `ZTF_ops` collection on `Kowalski` for new ZTF data (Twilight only by default).
- Uses `dask.distributed` to process individual ZTF image frames (ccd-quads). Each worker is initialized with a `TailsWorker` instance that maintains a `Fritz` connection and preloads Tails. The candidate comet detections, if any, are posted to `Fritz` together with auto-annotations (cross-matches from the MPC and SkyBot) and auxiliary data.

On Fritz, the posted results look like this:

- Candidates page
fritz-candidates

- Source page
fritz-source

This is an example of full-sized cutouts posted to Fritz:
ztf_20191014495961_000570_zr_c05_o_q3_0

### CODE REVIEW

**Comment by dmitryduev** [comment]
@stefanv @doccosmos green on CI and ready for review when you have a spare minute or two :)

**Comment by dmitryduev** [comment]
Testing out the service and looks like it works! See, for example:
https://fritz.science/source/ZTF20210113_urhb0g

**Review by stefanv — COMMENTED** [review]
@dmitryduev Impressive gluing together of all the various pieces!

There's a lot, so I did not try to evaluate whether every single line does what it's supposed to.

Generally, the code looks good and is tidy.  It would benefit from some consistency (docstrings everywhere, decide whether or not you want type annotation, use one type of documentation format, remove old comments, etc.).

What wasn't clear to me here is whether the tests do anything other than run the process.  It would be good to validate that the results are what you'd expect.  Maybe that is already in place and I just missed it?

Good job!

File: .github/workflows/test-run-tails.yaml [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        run: |
+          cd scripts
+          python run_tails.py run \
+            --config=../config.defaults.yaml \
+            --checkpoint=../models/tails-20210107/tails \
+            --score_threshold=0.5 \
+            --cleanup=none \
+            --single_image=ztf_20191014495961_000570_zr_c05_o_q3
+          ls runs/20191014
+          python -c "from pathlib import Path; assert len(list(Path('./runs/20191014').glob('*.*'))) == 3"
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply .github/workflows/test-run-tails.yaml:44]: That script runs Tails on an image that contains a comet. Strictly speaking, it is separate from the sentinel service -- I just added a test here to decrease the entropy of the Universe... 
The image is tessellated into 169 tiles, and on one of the them, the comet should be detected, which produces 3 files -- this is what I am checking here. You are right that it's probably a good idea to check what exactly the script produces, i.e. the comet position etc. Will write a dedicated test.

File: sentinel.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+    except Exception:
+        print(f"\r[✗] {message}")
+        raise
+    else:
+        print(f"\r[✓] {message}")
+
+
+def deps_ok():
+    """
+    Borrowed from https://github.com/fritz-marshal/fritz
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply sentinel.py:72]: Right, on Fritz it says the correct thing; fixed.

File: sentinel.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+RUN /opt/conda/bin/conda install -c conda-forge astromatic-swarp && \
+    ln -s /opt/conda/bin/swarp /bin/swarp && \
+    /opt/conda/bin/pip install -U pip && \
+    /opt/conda/bin/python setup.py install && \
+    /opt/conda/bin/pip install -r requirements.txt --no-cache-dir && \
+    /opt/conda/bin/python generate_supervisor_conf.py watcher
+
+# run container
+#RUN touch /app/1.txt
+#CMD tail -f 1.txt
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply sentinel.Dockerfile:38]: fixed;

File: sentinel.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+                except Exception:
+                    raise ValueError("Could not parse version")
+
+                if not (Version(version) >= Version(min_version)):
+                    raise RuntimeError(f"Required {min_version}, found {version}")
+        except ValueError:
+            print(
+                f"\n[!] Sorry, but our script could not parse the output of "
+                f'`{" ".join(cmd)}`; please file a bug, or see '
+                f"`check_app_environment.py`\n"
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply sentinel.py:96]: Removed.

File: sentinel.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+            "for installation instructions."
+        )
+        print()
+        return False
+
+    print("-" * 20)
+    return True
+
+
+def check_configs(config_wildcards=("config.*yaml", "docker-compose.*yaml")):
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply sentinel.py:124]: I do usually use one format, but here I wanted to try Typer, and my usual format caused it to misbehave, that's where the weird-looking docstring are coming from. Prob a good idea to just ditch `Typer` and use `fire` as I do everywhere else.

[the remark this answers is not in the document]
Comment (dmitryduev) [reply sentinel.py:114]: Replaced `Typer` with `fire` and fixed the docstrings.

File: ztf-sentinel-service/watcher.py [hunk 6]
```diff
@@ -0,0 +1,847 @@
...
+                f"{config['sentinel']['fritz']['protocol']}://"
+                f"{config['sentinel']['fritz']['host']}:{config['sentinel']['fritz']['port']}"
+                f"{endpoint}",
+                json=data,
+                headers=self.session_headers,
+            )
+
+        return response
+
+    def process_frame(self, frame):
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply ztf-sentinel-service/watcher.py:181]: Fixed

File: sentinel.py [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+        )
+
+        if not all(containers_up):
+            print("Sentinel's containers are not up, waiting...")
+            time.sleep(2)
+            continue
+
+        break
+
+    print("Testing sentinel")
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply sentinel.py]: Fixed

File: .github/workflows/test-run-tails.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dmitryduev) [reply .github/workflows/test-run-tails.yaml]: Improved readability and testing of the `run_tails.py` script.

**Comment by dmitryduev** [comment]
@stefanv I think I addressed your comments, please lmk if it looks good.

**Comment by dmitryduev** [comment]
@AshishMahabal @doccosmos just double checking: were you guys going to review this after yesterday? Or should I ask someone else to review?

**Comment by doccosmos** [comment]
Dima - I’m burnt out from TESS proposals this week but will review on Monday.

— Matthew

File: readme.md [hunk 8]
```diff
@@ -1,21 +1,76 @@
 # Tails: chasing comets with the Zwicky Transient Facility and Deep Learning
```
Comment (AshishMahabal) [inline readme.md:1]: May be "using Deep Learning"? Right now ZTF and DL could be construed to be on-par this way.

File: scripts/run_tails.py [hunk 9]
```diff
@@ -1,26 +1,23 @@
 #!/usr/bin/env python
-
-import argparse
 from astropy.coordinates import SkyCoord
 from astropy.time import Time
 from copy import deepcopy
 import datetime
-import matplotlib.pyplot as plt
```
Comment (AshishMahabal) [inline scripts/run_tails.py:8]: Were there plots before that have been removed at some point?

File: scripts/run_tails.py [hunk 10]
```diff
@@ -51,26 +54,36 @@ def fetch_data(_nsp):
...
 
 
 def process_ccd_quad(
-    name,
-    p_date,
-    checkpoint,
+    name: str,
+    p_date: pathlib.Path,
     model,
-    secrets,
```
Comment (AshishMahabal) [inline scripts/run_tails.py:63]: secrets not needed any more?

Comment (AshishMahabal) [reply scripts/run_tails.py:63]: I guess that is in config now.

File: scripts/run_tails.py (hunk 9 above)
Comment (AshishMahabal) [reply scripts/run_tails.py:8]: Doesn't plot_stack need it?

File: scripts/run_tails.py [hunk 11]
```diff
@@ -145,17 +157,13 @@ def process_ccd_quad(
...
                 reticles=((x_o, y_o),),
                 w=6,
                 h=2,
                 dpi=360,
                 save=str(p_date / f"{name}_{ni}.png"),
-                # cmap=cmr.arctic,
-                # cmap=plt.cm.viridis,
-                # cmap=plt.cm.cividis,
-                cmap=plt.cm.bone,
+                cmap="bone",
```
Comment (AshishMahabal) [inline scripts/run_tails.py:166]: I have not used bone, but it seems good. Any particular reasons for choosing it?

File: ztf-sentinel-service/generate_supervisor_conf.py [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    generate_conf("watcher")
+
+
+with open("/app/config.yaml", "r") as f:
+    config = yaml.load(f, Loader=yaml.FullLoader)
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser()
```
Comment (AshishMahabal) [inline ztf-sentinel-service/generate_supervisor_conf.py:30]: It is these kinds of things that have been replaced with fire elsewhere, right?

**Comment by AshishMahabal** [comment]
Hi Dima, Looks good overall. Made a few comments. Let us see if you need to make any changes there and then we should be good.

File: readme.md (hunk 8 above)
Comment (dmitryduev) [reply readme.md:1]: That's the intent! I'm foreseeing adding transfer learning examples for other surveys.

File: scripts/run_tails.py (hunk 9 above)
Comment (dmitryduev) [reply scripts/run_tails.py:8]: It does need it, but it's taken care of by the `from tails.utils import plot_stack` statement.

File: scripts/run_tails.py (hunk 10 above)
Comment (dmitryduev) [reply scripts/run_tails.py:63]: Yes, it's in the config.

File: scripts/run_tails.py (hunk 11 above)
Comment (dmitryduev) [reply scripts/run_tails.py:166]: Experimented with a ton of other (default) options over the years and like this one the most. @stefanv's inventions like `viridis` are great, but I prefer a more monochrome feel for single-band optical images. BTW, we use `bone` on Fritz by default.

File: ztf-sentinel-service/generate_supervisor_conf.py (hunk 12 above)
Comment (dmitryduev) [reply ztf-sentinel-service/generate_supervisor_conf.py]: Ah, right, forgot about this one. This script migrated from an old code base and I forgot to change it upon @stefanv's comment. Fixed now.

**Comment by dmitryduev** [comment]
@AshishMahabal thanks for the review! Addressed your comments.

File: ztf-sentinel-service/generate_supervisor_conf.py (hunk 12 above)
Comment (AshishMahabal) [reply ztf-sentinel-service/generate_supervisor_conf.py]: Has this change not been committed? I think I am still seeing the earlier version, and no diff.

**Review by AshishMahabal — APPROVED** [review]
Looks good. For some reason it did not show me diffs in one file, but I was able to see it after specifically going to commits.

**Comment by doccosmos** [comment]
I would add a bullet point describing the base dask configuration: two workers, etc., and something somewhere stating what the size of the Docker image works out as

**Comment by dmitryduev** [comment]
@doccosmos thanks! I added relevant text to `readme.md` and posted an issue regarding the Docker image size (#6). Feel free to merge if happy!

**Review by doccosmos — APPROVED** [review]

**Pull request merged**